src/opm/parser/eclipse/EclipseState/Grid/EclipseGrid.cpp

```diff
@@ -238,37 +352,145 @@ namespace Opm {
+        }
+
+        const std::vector<std::string>& gridunit = egridfile.get<std::string>("GRIDUNIT");
+        
+        const std::vector<int>& gridhead = egridfile.get<int>("GRIDHEAD");
+        const std::array<int, 3> dims = {gridhead[1], gridhead[2], gridhead[3]};
+
+        m_nx=dims[0];
+        m_ny=dims[1];
+        m_nz=dims[2];
+
+        const std::vector<float>& coord_f = egridfile.get<float>("COORD");
+        const std::vector<float>& zcorn_f = egridfile.get<float>("ZCORN");
+
+        m_coord.assign(coord_f.begin(), coord_f.end());
+        m_zcorn.assign(zcorn_f.begin(), zcorn_f.end());
+
+        if (gridunit[0] != "METRES") {
+
+            const auto length = ::Opm::UnitSystem::measure::length; 
+            
+            if (gridunit[0] == "FEET"){
+                Opm::UnitSystem units(Opm::UnitSystem::UnitType::UNIT_TYPE_FIELD );
+                units.to_si(length, m_coord);
+                units.to_si(length, m_zcorn);
+            } else if (gridunit[0] == "CM"){
+                Opm::UnitSystem units(Opm::UnitSystem::UnitType::UNIT_TYPE_LAB );
+                units.to_si(length, m_coord);
+                units.to_si(length, m_zcorn);
+            } else {
+                std::string message = "gridunit '" + gridunit[0] + "' doesn't correspong to a valid unit system";
+                throw std::invalid_argument(message);
+            }
+        }
+        
+        if ((egridfile.hasKey("ACTNUM")) && (m_useActnumFromGdfile)) {
+            const std::vector<int>& actnum  = egridfile.get<int>("ACTNUM");
+            resetACTNUM( actnum );
+        } 
+
+        if (egridfile.hasKey("MAPAXES")) {
+            const std::vector<float>& mapaxes_f = egridfile.get<float>("MAPAXES");
+            m_mapaxes.assign(mapaxes_f.begin(), mapaxes_f.end());
+        }
+
+        if (egridfile.hasKey("MAPUNITS")) {
+            const std::vector<std::string>& mapunits = egridfile.get<std::string>("MAPUNITS");
+            m_mapunits=mapunits[0];
+        }
+    
+        ZcornMapper mapper( getNX(), getNY(), getNZ());
+        zcorn_fixed = mapper.fixupZCORN( m_zcorn );
+    }
 
+    bool EclipseGrid::keywInputBeforeGdfile(const Deck& deck, const std::string keyword) const {
 
+        std::vector<std::string> keywordList; 
+        keywordList.reserve(deck.size());
+        
+        for (size_t n=0;n<deck.size();n++){
+            DeckKeyword kw = deck.getKeyword( n );
+            keywordList.push_back(kw.name());
+        }
+        
+        int indKeyw = -1;
+        int indGdfile = -1;
+        
+        for (size_t i = 0; i < keywordList.size(); i++){
+            if (keywordList[i]=="GDFILE"){
+                indGdfile=i;
+            }
+
+            if (keywordList[i]==keyword){
+                indKeyw=i;
+            }
+        }
+
+        if (indGdfile == -1) {
+            throw std::runtime_error("keyword GDFILE not found in deck ");
+        }
+
+        if (indKeyw == -1) {
```

Comment:
I would assume you could just return false in this case?

Reply:
I would prefer throw in this case. This function is also used in a context where return of false is interpreted as the keywords follows after not before.

Reply:
Sure - will not argue that, just a reflection: As I see it this an example of where the primary developer - i.e. you on this case, and the users (as in other developers) see and feel things differently.

As a developer you want 100% control and to be absolutely certain of stringent correctness, as a user I think your view would be more: "Ohh come on - give me a sensible default, don't be so f... cheap". Striking the balance is not easy - I have certainly erred on the strict myself.